fix #8940 feat(cirrus): Move objects into main file 🧹

yashikakhurana opened the pull request:
Because

- Python `parses` the file line by line and it may cause an issue when creating several objects like` fml, SDK,` etc, currently, we are creating objects into their individual class file and importing them into the `main`

This commit

- Deletes unnecessary files
- move objects into the main file and separate method for setup
- FML- Exception handling and exit the container` if the path or channel is invalid `
- SDK- Exception handling and exit the container if `context is missing any required argument`
- Removes file path from docker volumes for feature manifest

UPDATE: uses the concept of `lifespan` instead of using `startup and shutdown events`, thanks to @jaredlockhart 
- This PR also updates the version of Fast API to support `lifespan`
closes #8940

jrbenny35 commented on cirrus/server/tests/test_experiment_recipes.py:
```diff
@@ -4,11 +4,16 @@
+from cirrus.settings import context
 
 
 class TestFetchRecipes(unittest.TestCase):
```
Why unittest and not pytest?

jaredlockhart replied:
Usually it's prudent to avoid adding any additional dependencies until they're necessary.  Python has a robust builtin test framework that is used broadly across Nimbus, for instance by all of the Experimenter unit tests.  I wouldn't suggest incorporating pytest unless there's a specific need to.

yashikakhurana replied:
so currently unittest meets our testing requirements, and incorporating pytest would introduce an additional dependency without a specific need. However, we can keep pytest in mind for future considerations :)

jrbenny35 replied:
> Usually it's prudent to avoid adding any additional dependencies until they're necessary. Python has a robust builtin test framework that is used broadly across Nimbus, for instance by all of the Experimenter unit tests. I wouldn't suggest incorporating pytest unless there's a specific need to.

There's pytest being declared wthin some of these tests already. Also, a lot of the tests redefine the same variables for each test, pytest would solve this easily.  We have pytest for the Rust integration tests, now unit test here, and we will probably use pytest for the integration tests here. That seems like a lot of fragmentation for testing, when we could just use pytest everywhere and have a matching framework for all the tests.

> Sadly, the builtin methods in the Python unittest framework use camelcase.

No need for setup functions either with pytest.

jaredlockhart replied:
@jrbenny35 Yes super duper good points, we should absolutely be consistent in which patterns and frameworks we're using.  If we're already using some pytest constructs we should continue to, 100% agree 👍

yashikakhurana replied:
yeah, my mistake, I missed that, I think in starting tests I added, then started using unittest, this also requires some changes in other test files too- so added a ticket, will do follow up in here- https://github.com/mozilla/experimenter/issues/8993

jrbenny35 commented on cirrus/server/tests/test_experiment_recipes.py:
```diff
@@ -4,11 +4,16 @@
 
 
 class TestFetchRecipes(unittest.TestCase):
+    def setUp(self):
+        self.sdk = SDK(context)
```
a wild Camel appears 
```suggestion
    def set_up(self):
        self.sdk = SDK(context)
```

jaredlockhart replied:
Sadly, the builtin methods in the Python unittest framework use [camelcase](https://docs.python.org/3/library/unittest.html#unittest.TestCase.setUp).

yashikakhurana replied:
I completely understand @jrbenny35 how weird it looks in python code, but builtin methods :(

jrbenny35 commented on cirrus/server/tests/test_main.py:
```diff
@@ -1,6 +1,51 @@
+from cirrus.sdk import SDK
+
+
+class TestSetup(unittest.TestCase):
+    def test_setup_with_fml_error(self):
```
These really should be using pytest, no reason to use unittest just for cirrus.

jrbenny35 reviewed (changes requested):
Curious as to why we are using unittest over pytest

jaredlockhart commented on cirrus/server/tests/test_feature_manifest.py:
```diff
@@ -6,12 +6,23 @@
+        self.fml = FeatureManifestLanguage(fml_path, channel)
+        self.sdk = SDK(context=context)
+
+    def test_invalid_fml_path(self):
```
👏 👏 👏

jaredlockhart commented on cirrus/server/tests/test_feature_manifest.py:
```diff
@@ -6,12 +6,23 @@
+            FeatureManifestLanguage("invalid_fml_path.txt", channel)
+        self.assertTrue(str(e.exception).startswith("FmlError.InvalidPath"))
+
+    def test_invalid_channel(self):
```
👍 👍 👍

jaredlockhart commented on cirrus/server/tests/test_sdk.py:
```diff
@@ -1,13 +1,52 @@
 from cirrus_sdk import NimbusError
+from parameterized import parameterized
 
+from cirrus.main import sdk
```
This is the same pattern again.  Don't initialize stateful objects in the global namespace at all.  If you need an instance of `SDK` in the tests, initialize it inside the `setUp` method.

jaredlockhart commented on cirrus/server/cirrus/main.py:
```diff
@@ -1,23 +1,63 @@
+    return app, sdk, remote_setting, fml, scheduler
+
+
+app, sdk, remote_setting, fml, scheduler = setup()
```
This is the same problem.  Don't initialize stateful objects in the global namespace, it looks like FastAPI uses a [dependency injection](https://fastapi.tiangolo.com/tutorial/dependencies/#declare-the-dependency-in-the-dependant) style.  Give that a shot.

yashikakhurana replied:
good find, actually this got introduced in version `0.93.0` and I was using `0.92.0 `(that time it was current :D), so I updated the fast API to a latest version now- `0.97.0` in this PR, this is actually neat, no need to have a separate method for startup and shutdown, thanks @jaredlockhart 🌷

jaredlockhart reviewed (changes requested):
Getting there.  Left a couple pointers for main.py and tests.

brennie commented on cirrus/server/cirrus/sdk.py:
```diff
@@ -4,14 +4,12 @@
-    def __init__(self):
-        self.client = CirrusClient(context)
+    def __init__(self, context: str):
+        self.client = CirrusClient(context)  # does not match struct- throw an exception
```
What does this comment mean?

yashikakhurana replied:
so CirrusClient is actually coming from nimbus_sdk, if the context we are providing doesn't match what it is expecting, it will throw an error

brennie commented:
Could you clarify your summary/description with regards to "move objects into main file"? That isn't super clear -- what are the objects in question?

yashikakhurana commented:
> Could you clarify your summary/description with regards to "move objects into main file"? That isn't super clear -- what are the objects in question?

Absolutely- The reason behind consolidating the object creation into the `main.py` file was to address potential issues that can arise when creating multiple objects in separate files. Python parses files line by line, and if the objects are created in their individual class files, there is a possibility of encountering dependency or import-related problems. By moving the object creation process to the main file and importing the necessary classes, just making ensure that the objects are created in the correct order and any dependencies are properly handled. so that to avoid potential runtime errors, let me know if it helps :)

jaredlockhart reviewed (approved):
Okay cool!  This is starting to look really good.  In terms of addressing the issues of stateful globals, I think this looks good.  @jrbenny35 raises some good points about fragmentation in testing patterns around unittest and pytest that I think would be good to think about but can probably be done as a followup rather than blowing up the scope of this pr.  R+ from me 👍

yashikakhurana commented:
> Okay cool! This is starting to look really good. In terms of addressing the issues of stateful globals, I think this looks good. @jrbenny35 raises some good points about fragmentation in testing patterns around unittest and pytest that I think would be good to think about but can probably be done as a followup rather than blowing up the scope of this pr. R+ from me 👍

yup absolutely, @jrbenny35 thank you for noticing that, @jaredlockhart I added a follow up ticket- https://github.com/mozilla/experimenter/issues/8993
